### DISCUSSION TITLE
Problems with MariaDB running on the same port as MySQL and missing nginx configuration.

### DISCUSSION THREAD

**Discussion by creative-andrew** [postalserver/postal#1534] [Help with using Postal] [answered]
Hi there! 

I am installing postal on my VPS and I am running with the following issues.

Please note that my VPS has already software installed, including MySQL.

(1) When adding the MariaDB docker image, 

I am getting this error: 

```
ubuntu@app:/home$ sudo docker start postal-mariadb
Error response from daemon: driver failed programming external connectivity on endpoint postal-mariadb (ccad082db4aece666e179f26a6a795f49af5393a2a638ad86878984ce1c7b903): Error starting userland proxy: listen tcp4 127.0.0.1:3306: bind: address already in use
Error: failed to start containers: postal-mariadb
```
How can I make this run in a different port? 

(2) There is no information on how to configure Nginx to launch postal. Which config do I need to put? Where is the root folder of the application?

I remember that with Postal 1 we used to do something like this:

```
server {
    listen [::]:443 ssl default_server;
    listen 0.0.0.0:443 ssl default_server;
    root /opt/postal/app/public;
    server_name smtp.emailsender.club;
    ssl_certificate      ssl/postal.cert;
    ssl_certificate_key  ssl/postal.key;

    # Generate using: openssl dhparam 4096 -out /etc/ssl/dhparam.pem
    # ssl_dhparam /etc/ssl/dhparam.pem;

    ssl_protocols TLSv1.2 TLSv1.1 TLSv1;
    ssl_prefer_server_ciphers on;
    ssl_ciphers EECDH+ECDSA+AESGCM:EECDH+aRSA+AESGCM:EECDH+ECDSA+SHA512:EECDH+ECDSA+SHA384:EECDH+ECDSA+SHA256:ECDH+AESGCM:ECDH+AES256:DH+AESGCM:DH+AES256:RSA+AESG$

    location / {
       client_max_body_size 50M;
       try_files $uri $uri/index.html $uri.html @puma;
    }

    location /assets {
       add_header Cache-Control max-age=3600;
    }

    location @puma {
        proxy_set_header X-Real-IP  $remote_addr;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
        proxy_pass http://127.0.0.1:5000;
    }
}
```
Thanks a lot!!!!

Regards,
A.

**Comment by adamcooke** [comment, marked as answer]
Firstly, we highly recommend installing Postal on a dedicated machine. Any other configuration is not officially supported so you'll need to do a bit of research with Docker if you do want to use it to run another database instance on the same host. If you haven't used it yet, you can just delete `postal-mariadb` and make it again and change the local port in the `docker run` command (`127.0.0.1:3307:3306`). You don't need a separate database server though, you can just use the one you already have running and put its credentials in your Postal config.

For nginx, the only thing you need to do is forward all traffic to `127.0.0.1:5000`. There is no need to specify a root directory although you will need to configure SSL as appropriate.

**Reply by creative-andrew** [reply]
Thanks a lot for your reply:

I used:

```
ubuntu@app:/home$ sudo docker run -d \
   --name postal-mariadb \
   -p 127.0.0.1:3307:3306 \
   --restart always \
   -e MARIADB_DATABASE=postal \
   -e MARIADB_ROOT_PASSWORD=postal \
   mariadb
dbafc09c9a9cce7be1b5b1f2e7abbe309d856e4e80d2fc4e74a8c12770ef21d0
```
And it went well, but when using postal initialize, I get the following error:

ubuntu@app:/home$ sudo postal initialize

```
Pulling web      ... done
Pulling smtp     ... done
Pulling worker   ... done
Pulling cron     ... done
Pulling requeuer ... done
Creating postal_runner_run ... done
Initializing database
rake aborted!
URI::InvalidURIError: the scheme mysql2 does not accept registry part: root:postal@127.0.0.1:3307:3306 (or bad hostname?)
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:40:in `initialize'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:282:in `new'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:282:in `resolve_url_connection'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:270:in `resolve_hash_connection'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:245:in `resolve_connection'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:143:in `resolve'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:163:in `block in resolve_all'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:162:in `each'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_adapters/connection_specification.rb:162:in `resolve_all'
/usr/local/bundle/gems/activerecord-5.2.6/lib/active_record/connection_handling.rb:72:in `resolve'
```
I have modified the posta.yml file with several values (no port, 3307, 3307:3306 with no luck. 

```
main_db:
  # Specify the connection details for your MySQL database
  host: 127.0.0.1:3307
  username: root
  password: postal
  database: postal

message_db:
  # Specify the connection details for your MySQL server that will be house the
  # message databases for mail servers.
  host: 127.0.0.1:3307
  username: root
  password: postal
  prefix: postal
```

```
ubuntu@app:/home$ sudo docker ps
CONTAINER ID   IMAGE                               COMMAND                  CREATED             STATUS             PORTS                                                                      NAMES
dbafc09c9a9c   mariadb                             "docker-entrypoint.s…"   12 minutes ago      Up 12 minutes      127.0.0.1:3307->3306/tcp  
```

**Reply by creative-andrew** [reply]
I think it's not using MariaDB but going directly to MySQL.

**Reply by willpower232** [reply]
Have a look at https://github.com/postalserver/postal/blob/main/config/postal.defaults.yml, the port must be specified as a different line to the host

**Reply by creative-andrew** [reply]
@willpower232 it worked perfectly! thanks like always!

**Comment by creative-andrew** [comment]
@adamcooke  @willpower232 

I have solved the problem regarding MariaDB and Mysql. The proposed solution worked great.

Based on your recommendation I have forwarded traffic to 127.0.0.1:5000 using the following configuration

```
server {
        server_name yourdomaim.com www.yourdomaim.com;
        listen IP;
        location / {
                proxy_pass "http://127.0.0.1:5000/";
        }
        listen IP:443 ssl;
        ssl_certificate /home/smtp/ssl.combined;
        ssl_certificate_key /home/smtp/ssl.key;
}
```
and it loads correctly.

However, when I try to access with my password, I get the following message:

```
The change you wanted was rejected.
Maybe you tried to change something you didn't have access to.

If you are the application owner check the logs for more information.
```

My postal status looks like this:

```
      Name                     Command               State    Ports
-------------------------------------------------------------------
postal_cron_1       /docker-entrypoint.sh post ...   Up            
postal_requeuer_1   /docker-entrypoint.sh post ...   Up            
postal_smtp_1       /docker-entrypoint.sh post ...   Exit 1        
postal_web_1        /docker-entrypoint.sh post ...   Up            
postal_worker_1     /docker-entrypoint.sh post ...   Up   
```

I have checked the logs, but I don't find anything in particular. 

Here is a fragment of the logs when I interact with the webapp: https://pastebin.com/DnkJntYW

Might this error be related to the fact that postal_smtp_1 has Exit 1?

Thanks a lot!

**Reply by creative-andrew** [reply]
Everything working now.

The problem was the missing headers:

the final Nginx. configuration ended as following:

```
server {
        listen YOUR_IP:80;
        server_name yourdomaim.com www.yourdomaim.com;
        return 301 https://$host$request_uri;
}

server {
    listen YOURIP:443 ssl;
    server_name yourdomaim.com www.yourdomaim.com;
    ssl_certificate PATH_OF_SSL
    ssl_certificate_key PATH_OF_SSL;

    location / {
        add_header  X-Robots-Tag "noindex, nofollow, nosnippet, noarchive";
        proxy_set_header X-Real-IP  $remote_addr;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
        proxy_pass http://127.0.0.1:5000;
    }
}
```
Maybe this can help someone as it's a really basic config.

**Reply by willpower232** [reply]
My notes also have `proxy_set_header X-Forwarded-Proto https;` so that may or may not really be needed but better to set them all I guess.

You may also like `add_header X-Robots-Tag "noindex, nofollow, nosnippet, noarchive"` to make sure you aren't accidentally indexed.

Also the status you posted suggests your SMTP is not working which may or may not be something you were aware of.

**Reply by willpower232** [reply]
Sorry I didn't see you'd already mentioned it, as the SMTP is still listed you should be able to access its logs as well with `postal logs smtp`.

**Reply by creative-andrew** [reply]
@willpower232 

Thanks for your suggestions. I will add the additional config to Nginx. 

Apparently, the Port 25 is already in use:

```
Attaching to postal_smtp_1
smtp_1      | rake aborted!
smtp_1      | 
smtp_1      | Errno::EADDRINUSE: Address already in use - bind(2) for "::" port 25
smtp_1      | 
smtp_1      | /opt/postal/app/lib/postal/smtp_server/server.rb:52:in `initialize'
smtp_1      | /opt/postal/app/lib/postal/smtp_server/server.rb:52:in `open'
smtp_1      | /opt/postal/app/lib/postal/smtp_server/server.rb:52:in `listen'
smtp_1      | /opt/postal/app/lib/postal/smtp_server/server.rb:259:in `run'
smtp_1      | /opt/postal/app/lib/tasks/postal.rake:13:in `block (2 levels) in <top (required)>'
smtp_1      | /opt/postal/app/bin/bundle:3:in `load'
smtp_1      | /opt/postal/app/bin/bundle:3:in `<main>'
smtp_1      | 
smtp_1      | Tasks: TOP => postal:smtp_server
smtp_1      | 
smtp_1      | (See full trace by running task with --trace)
smtp_1      | 
```

What would be the normal workaround when 2 apps need to listen to a single port?

**Reply by willpower232** [reply]
Turn one of them off :sweat_smile: 

It might be that your OS came with an email server like exim or postfix pre-installed so that could be removed without too much hassle.

If you are running two email services (Postal and something else) on the same server, one will need to be moved.

**Reply by creative-andrew** [reply]
@willpower232 

I turned postfix off and it worked.

However, I'm running Web/Virtualmin with some WordPress installations running. I wonder if the native mail() function will work now or if it depends on postfix. Not really sure. A lot of moving pieces here haha.

Thanks like always. You are a gentleman and a scholar.

**Reply by willpower232** [reply]
We do not recommend running Postal on the same server as other things so if you can get a second server, it would be much easier for you.

You will need to reconfigure each site to create an authenticated email flow for them as Postal does not allow emails to just be sent without authentication.

**Reply by creative-andrew** [reply]
Yeah, I will end up doing that. That was just me being cheap.
